explorer/ast/declaration.h

```diff
@@ -102,38 +102,22 @@ class Declaration : public AstNode {
   std::optional<Nonnull<const Value*>> constant_value_;
 };
 
-class FunctionDeclaration : public Declaration {
+class CallableDeclaration : public Declaration {
  public:
-  using ImplementsCarbonValueNode = void;
-
-  static auto Create(Nonnull<Arena*> arena, SourceLocation source_loc,
-                     std::string name,
-                     std::vector<Nonnull<AstNode*>> deduced_params,
-                     std::optional<Nonnull<Pattern*>> me_pattern,
-                     Nonnull<TuplePattern*> param_pattern,
-                     ReturnTerm return_term,
-                     std::optional<Nonnull<Block*>> body)
-      -> ErrorOr<Nonnull<FunctionDeclaration*>>;
-
-  // Use `Create()` instead. This is public only so Arena::New() can call it.
-  FunctionDeclaration(SourceLocation source_loc, std::string name,
+  CallableDeclaration(AstNodeKind kind, SourceLocation loc, std::string name,
                       std::vector<Nonnull<GenericBinding*>> deduced_params,
                       std::optional<Nonnull<Pattern*>> me_pattern,
                       Nonnull<TuplePattern*> param_pattern,
                       ReturnTerm return_term,
                       std::optional<Nonnull<Block*>> body)
-      : Declaration(AstNodeKind::FunctionDeclaration, source_loc),
+      : Declaration(kind, loc),
         name_(std::move(name)),
         deduced_parameters_(std::move(deduced_params)),
         me_pattern_(me_pattern),
         param_pattern_(param_pattern),
         return_term_(return_term),
         body_(body) {}
 
-  static auto classof(const AstNode* node) -> bool {
-    return InheritsFromFunctionDeclaration(node->kind());
-  }
-
   void PrintDepth(int depth, llvm::raw_ostream& out) const;
 
   auto name() const -> const std::string& { return name_; }
```

Comment:
I think not! I want to avoid that we have to cast unnecessarily in DeclareCallableDeclaration.
Is that ok with you?

Reply:
I'm concerned here mainly because providing a `name()` for destructors encourages accidents involving treating `destructor` as a function name. Removing the function here would avoid that.

Looking at DeclareCallableDeclaration, where it does:

`**trace_stream_ << "** declaring function " << f->name() << "\n";`

To avoid a cast, that could instead be:

```
**trace_stream_ << "** declaring function ";
f->PrintID(**trace_stream);
**trace_stream_ << "\n";
```

The one spot I see `name()` used is for determining whether the function can be an entry point. A cast probably makes sense for checking if something is an entry point; destructors shouldn't be valid entry points.

Reply:
Sorry, name() would be used in value.cpp resolve_names.cpp and interpreter.cpp...

Reply:
I expect it to require code changes, but it's probably the right direction to head in: if you don't want to do this here, can you please add a TODO so that we don't forget about it?